lib/upload_to_s3.js

```diff
@@ -1,45 +1,84 @@
-      // be used with azure simply by changing s3 -> azure.
-      storageType: 's3',
-      expires: new Date(Math.min(expiration, new Date(task.task.expires))),
-      contentType: httpsHeaders['content-type']
-    }
-  );
-
-  let parsedUrl = url.parse(artifact.putUrl);
-  let options = {
-    hostname: parsedUrl.hostname,
-    path: parsedUrl.path,
-    method: 'PUT',
-    headers: httpsHeaders
-  };
-
-  let req = https.request(options);
-  sourceStream.pipe(req);
-
-  let response = await new Promise((accept, reject) => {
-    req.on('response', accept);
-    req.on('error', reject);
-    req.setTimeout(5 * 60 * 1000, reject);
+export default async function uploadToS3 (
+  task,
+  sourceStream,
+  artifactName,
+  expiration,
+  httpsHeaders,
+  putUrl,
+  additionalOptions)
+{
+  let queue = task.runtime.queue;
+
+  let tmp = new temporary.File();
+  await new Promise((accept, reject) => {
+    let stream = fs.createWriteStream(tmp.path);
+    stream.on('error', reject);
+    stream.on('finish', accept);
+    sourceStream.pipe(stream);
   });
 
-  // Flush the data from the reponse so it's not held in memory
-  response.resume();
+  try {
+    if (!putUrl) {
+      var artifact = await queue.createArtifact(
+        task.status.taskId,
+        task.runId,
+        artifactName,
+        {
+          // Why s3? It's currently cheaper to store data in s3 this could easily
+          // be used with azure simply by changing s3 -> azure.
+          storageType: 's3',
+          expires: new Date(Math.min(expiration, new Date(task.task.expires))),
+          contentType: httpsHeaders['content-type']
+        }
+      );
+
+      putUrl = artifact.putUrl;
+    }
+
+    var parsedUrl = url.parse(putUrl);
+    var options = extend({
+      hostname: parsedUrl.hostname,
+      path: parsedUrl.path,
+      method: 'PUT',
+      headers: httpsHeaders,
+      port: parsedUrl.port
+    }, additionalOptions);
+
+    await promiseRetry((retry, number) => {
+      if (number > 1) {
+        task.runtime.log('retrying artifact upload');
+      }
+
+      return new Promise((accept, reject) => {
+        let req = https.request(options);
+
+        req.on('response', (response) => {
+          // Flush the data from the reponse so it's not held in memory
+          response.resume();
+
+          if (response.statusCode !== 200) {
+            reject(new Error(
+              `Could not upload artifact. Status Code: ${response.statusCode}`
+            ));
+          } else {
+            accept();
+          }
+        });
 
-  if (response.statusCode !== 200) {
-    throw new Error(
-      `Could not upload artifact. ${response.error} Status Code: ${response.statusCode}`
-    );
+        req.on('error', reject);
+        req.setTimeout(5 * 60 * 1000, reject);
+        task.runtime.log(`Uploading ${artifactName}`);
+        fs.createReadStream(tmp.path).pipe(req);
+      }).catch(retry);
+    }, {randomize: true});
+  } finally {
+    tmp.unlink();
```

Comment:
you should handle errors here and at least log them... I'm not sure it's worth crashing over :)